README.md
# ClusterPrePermissions

Cluster's reusable pre-permissions utility that lets developers ask the users on their own dialog for photos or contacts access, before making the system-based request. This is based on [Cluster's post *The Right Way to Ask Users for iOS Permissions*](https://medium.com/p/96fa4eb54f2c).

# Requirements
* from version 0.1.8 on, please use Xcode 6, before 0.1.7(include), you can use Xcode 5
* ARC


# Why use Pre-Permissions?

On iOS, when your code tries to access the Camera Roll or the Address Book, the system automatically shows an alert to the user.

If the user says "Don't Allow" to this alert, **you have burned your one-and-only chance to get that access**. The user will have to jump through a 5-step process outside of your app to re-enable access, which they will rarely, if ever, do.

Often the user wasn't event expecting the access alert, and by habit will tap "Don't Allow". This is a lose-lose scenario for both you and your user.

At [Cluster](http://cluster.co), we’ve had a lot of success asking for permissions using our own UI before popping the iOS dialogs.

**For photos alone, 46% of people who denied access in a pre-permissions dialog ended up granting access when asked at a better time later.**

# How Pre-Permissions Works


This utility simply wraps your code with a "pre-permission" alert, if needed. When you need to ask for photo access, for example, this utility essentially just asks them twice.

![image](http://f.cl.ly/items/2I1V1R3b3q3A3H3y3u18/new-1.jpg)

Even though it seems annoying to ask twice, at Cluster we almost completely eliminated the possibility of a user tapping “Don’t Allow”, leaving the door open for us to win them back in the future. And for what it’s worth, in [live user tests](https://medium.com/ux-ui-design/9b7e9edd2de8), not a single test subject hesitated (or expressed confusion) when the second dialog appeared.

# Adding ClusterPrePermissions to your project

## Cocoapods

[CocoaPods](http://cocoapods.org/) is the recommended way to add ClusterPrePermissions to your project.

1. Add a pod entry for ClusterPrePermissions to your Podfile `pod 'ClusterPrePermissions', '~> 0.1'`
2. Install the pod(s) by running `pod install`
3. Include ClusterPrePermissions wherever you need it with `#import <ClusterPrePermissions/ClusterPrePermissions.h>`

## Source Files

Add `ClusterPrePermissions.h` and `ClusterPrePermissions.m` into your project.

1. Download the latest code, using `git clone`.
2. Open your project in Xcode, then drag and drop `ClusterPrePermissions.h` and `ClusterPrePermissions.m`onto your project (Make sure to select Copy items when asked if you extracted the code archive outside of your project.)
3. In your classes where you need ClusterPrePermissions, add it with `#import "ClusterPrePermissions.h"`


# Usage

CLPermissions checks the authorization status of photos and contacts, and, given a block callback, will return with the result. If authorization status is "not determined", the utility will show a configurable `UIAlertView` that asks for "pre-permission" before making the actual system request.

If the authorization status already unchangeable, your block will just be called back. You can check the `userDialogResult` and `systemDialogResult` to see what happened. If both are `ClusterDialogResultNoActionTaken`, then it means the access was already denied or granted by the system.

```objc
ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
[permissions showPhotoPermissionsWithTitle:@"Access your photos?"
                                   message:@"Your message here"
                           denyButtonTitle:@"Not Now"
                          grantButtonTitle:@"Give Access"
                         completionHandler:^(BOOL hasPermission,
                                             ClusterDialogResult userDialogResult,
                                             ClusterDialogResult systemDialogResult) {
                             if (hasPermission) {
                                 // Continue with your code here
                             } else {
                             	  // Handle access not being available
                             }
                         }];
```


CHANGELOG.md
# ClusterPrePermissions CHANGELOG

## 0.1.8
compatible for iOS8 push notification API changes.

## 0.1.7

support iOS5.0 to install this pod.

## 0.1.0

Initial release.


ClusterPrePermissions/CLAppDelegate.m
//
//  CLAppDelegate.m
//  ClusterPrePermissions
//
//  Created by Rizwan Sattar on 4/7/14.
//  Copyright (c) 2014 Cluster Labs, Inc. All rights reserved.
//

#import "CLAppDelegate.h"
#import "ClusterPrePermissions.h"
#import "PermissionsTestViewController.h"

@implementation CLAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
    // Override point for customization after application launch.
    self.window.backgroundColor = [UIColor whiteColor];
    self.window.rootViewController = [[PermissionsTestViewController alloc] initWithNibName:nil bundle:nil];
    [self.window makeKeyAndVisible];

    return YES;

}

-(void)application:(UIApplication*) application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
{
    BOOL didRegisterPushNotification = [ClusterPrePermissions didRegisterPushNotification];
    if(!didRegisterPushNotification) {
        [ClusterPrePermissions setResultForRigisterPushNotification:YES];
    }
    //Register device token stuff with whatever push notification system you're using
}

-(void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo
{
    //Handle information from remote notification here
    NSLog(@"%s, userinfo:%@", __func__, userInfo);
}

-(void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
    NSLog(@"%s, %@", __func__, error);
}

- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


ClusterPrePermissions/ClusterPrePermissions/ClusterPrePermissions.h
//
//  ClusterPrePermissions.h
//  ClusterPrePermissions
//
//  Created by Rizwan Sattar on 4/7/14.
//  Copyright (c) 2014 Cluster Labs, Inc. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

#import <Foundation/Foundation.h>

@interface ClusterPrePermissions : NSObject

/**
 * A general descriptor for the possible outcomes of a dialog.
 */
typedef NS_ENUM(NSInteger, ClusterDialogResult) {
    /// User was not given the chance to take action.
    /// This can happen if the permission was
    /// already granted, denied, or restricted.
    ClusterDialogResultNoActionTaken,
    /// User declined access in the user dialog or system dialog.
    ClusterDialogResultDenied,
    /// User granted access in the user dialog or system dialog.
    ClusterDialogResultGranted,
    /// The iOS parental permissions prevented access.
    /// This outcome would only happen on the system dialog.
    ClusterDialogResultParentallyRestricted
};

//Enum to cover the various states of push notification status authorization
typedef enum {
    kPushAuthorizationStatusAuthorized = 0,
    kPushAuthorizationStatusDenied,
    kPushAuthorizationStatusNotDetermined
} PushAuthorizationStatus;

/**
 * General callback for permissions. 
 * @param hasPermission Returns YES if system permission was granted 
 *                      or is already available, NO otherwise.
 * @param userDialogResult Describes whether the user granted/denied access, 
 *                         or if the user didn't have an opportunity to take action. 
 *                         ClusterDialogResultParentallyRestricted is never returned.
 * @param systemDialogResult Describes whether the user granted/denied access, 
 *                           or was parentally restricted, or if the user didn't 
 *                           have an opportunity to take action.
 * @see ClusterDialogResult
 */
typedef void (^ClusterPrePermissionCompletionHandler)(BOOL hasPermission,
                                              ClusterDialogResult userDialogResult,
                                              ClusterDialogResult systemDialogResult);

+ (instancetype) sharedPermissions;
+ (void)setResultForRigisterPushNotification:(BOOL)didRegisterPushNotification;
+ (BOOL)didRegisterPushNotification;
+ (PushAuthorizationStatus)pushAuthorizationStatus;

- (void) showPhotoPermissionsWithTitle:(NSString *)requestTitle
                               message:(NSString *)message
                       denyButtonTitle:(NSString *)denyButtonTitle
                      grantButtonTitle:(NSString *)grantButtonTitle
                     completionHandler:(ClusterPrePermissionCompletionHandler)completionHandler;

- (void) showContactsPermissionsWithTitle:(NSString *)requestTitle
                                  message:(NSString *)message
                          denyButtonTitle:(NSString *)denyButtonTitle
                         grantButtonTitle:(NSString *)grantButtonTitle
                        completionHandler:(ClusterPrePermissionCompletionHandler)completionHandler;

- (void) showLocationPermissionsWithTitle:(NSString *)requestTitle
                                  message:(NSString *)message
                          denyButtonTitle:(NSString *)denyButtonTitle
                         grantButtonTitle:(NSString *)grantButtonTitle
                        completionHandler:(ClusterPrePermissionCompletionHandler)completionHandler;

- (void) showPushNotificationPermissionsWithTitle:(NSString *)requestTitle
                                          message:(NSString *)message
                                  denyButtonTitle:(NSString *)denyButtonTitle
                                 grantButtonTitle:(NSString *)grantButtonTitle
                                completionHandler:(ClusterPrePermissionCompletionHandler)completionHandler;

//iOS 7.0+
- (void) showVideoCapturePermissionsWithTitle:(NSString *)requestTitle
                                      message:(NSString *)message
                              denyButtonTitle:(NSString *)denyButtonTitle
                             grantButtonTitle:(NSString *)grantButtonTitle
                            completionHandler:(ClusterPrePermissionCompletionHandler)completionHandler;

@end


ClusterPrePermissions/PermissionsTestViewController.m
//
//  PermissionsTestViewController.m
//  ClusterPrePermissions
//
//  Created by Rizwan Sattar on 4/7/14.
//  Copyright (c) 2014 Cluster Labs, Inc. All rights reserved.
//

#import "PermissionsTestViewController.h"

#import "ClusterPrePermissions.h"

@interface PermissionsTestViewController ()
@property (strong, nonatomic) IBOutlet UILabel *photoPermissionResultLabel;
@property (strong, nonatomic) IBOutlet UILabel *contactsPermissionResultLabel;
@property (strong, nonatomic) IBOutlet UILabel *locationPermissionResultLabel;
@property (strong, nonatomic) IBOutlet UILabel *pushNotificationPermissionResultLabel;
@property (weak, nonatomic) IBOutlet UILabel *videoCapturePermissionResultLabel;

@end

@implementation PermissionsTestViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (BOOL)prefersStatusBarHidden
{
    return YES;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)onPhotoPermissionsButtonTapped:(id)sender
{
    ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
    [permissions showPhotoPermissionsWithTitle:@"Access your photos?"
                                       message:@"Your message here"
                               denyButtonTitle:@"Not Now"
                              grantButtonTitle:@"Give Access"
                             completionHandler:^(BOOL hasPermission,
                                                 ClusterDialogResult userDialogResult,
                                                 ClusterDialogResult systemDialogResult) {
                                 [self updateResultLabel:self.photoPermissionResultLabel
                                          withPermission:hasPermission
                                        userDialogResult:userDialogResult
                                      systemDialogResult:systemDialogResult];
                             }];
}

- (IBAction)onContactsButtonPermissionTapped:(id)sender
{
    ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
    [permissions showContactsPermissionsWithTitle:@"Access your contacts?"
                                          message:@"Your message here"
                                  denyButtonTitle:@"Not Now"
                                 grantButtonTitle:@"Give Access"
                                completionHandler:^(BOOL hasPermission,
                                                    ClusterDialogResult userDialogResult,
                                                    ClusterDialogResult systemDialogResult) {
                                    [self updateResultLabel:self.contactsPermissionResultLabel
                                             withPermission:hasPermission
                                           userDialogResult:userDialogResult
                                         systemDialogResult:systemDialogResult];
                                }];
}

- (IBAction)onLocationButtonPermissionTapped:(id)sender
{
    ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
    [permissions showLocationPermissionsWithTitle:@"Access your location?"
                                          message:@"Your message here"
                                  denyButtonTitle:@"Not Now"
                                 grantButtonTitle:@"Give Access"
                                completionHandler:^(BOOL hasPermission,
                                                    ClusterDialogResult userDialogResult,
                                                    ClusterDialogResult systemDialogResult) {
                                    [self updateResultLabel:self.locationPermissionResultLabel
                                             withPermission:hasPermission
                                           userDialogResult:userDialogResult
                                         systemDialogResult:systemDialogResult];
                                }];
}

- (IBAction)onPushNotificationButtonTapped:(id)sender
{
    ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
    [permissions showPushNotificationPermissionsWithTitle:@"Allow Push Notifications?"
                                                  message:@"Your message here"
                                          denyButtonTitle:@"Not now"
                                         grantButtonTitle:@"Give access"
                                        completionHandler:^(BOOL hasPermission,
                                                            ClusterDialogResult userDialogResult,
                                                            ClusterDialogResult systemDialogResult) {
                                            [self updateResultLabel:self.pushNotificationPermissionResultLabel
                                                     withPermission:hasPermission
                                                   userDialogResult:userDialogResult
                                                 systemDialogResult:systemDialogResult];
                                        }];
}

- (IBAction)onVideoCapturePermissionsButtonTapped:(id)sender
{
    ClusterPrePermissions *permissions = [ClusterPrePermissions sharedPermissions];
    [permissions showVideoCapturePermissionsWithTitle:@"Allow Video Capture Permission?"
                                              message:@"Your message here"
                                      denyButtonTitle:@"Not now"
                                     grantButtonTitle:@"Give access"
                                    completionHandler:^(BOOL hasPermission,
                                                        ClusterDialogResult userDialogResult,
                                                        ClusterDialogResult systemDialogResult) {
                                        [self updateResultLabel:self.videoCapturePermissionResultLabel
                                                 withPermission:hasPermission
                                               userDialogResult:userDialogResult
                                             systemDialogResult:systemDialogResult];
                                    }];
}

- (void) updateResultLabel:(UILabel *)resultLabel
            withPermission:(BOOL)hasPermission
          userDialogResult:(ClusterDialogResult)userDialogResult
        systemDialogResult:(ClusterDialogResult)systemDialogResult
{
    resultLabel.text = @"haha";
    resultLabel.alpha = 0.0;

    if (hasPermission) {
        resultLabel.textColor = [UIColor colorWithRed:0.1 green:1.0 blue:0.1 alpha:1.0];
    } else {
        resultLabel.textColor = [UIColor colorWithRed:1.0 green:0.1 blue:0.1 alpha:1.0];
    }
    NSString *text = nil;
    if (userDialogResult == ClusterDialogResultNoActionTaken &&
        systemDialogResult == ClusterDialogResultNoActionTaken) {
        NSString *prefix = nil;
        if (hasPermission) {
            prefix = @"Granted.";
        } else if (systemDialogResult == ClusterDialogResultParentallyRestricted) {
            prefix = @"Restricted.";
        } else {
            prefix = @"Denied.";
        }
        text = [NSString stringWithFormat:@"%@ Dialogs not shown, system choice already made.", prefix];
    } else {
        NSString *userResultString = [self stringFromDialogResult:userDialogResult];
        NSString *systemResultString = [self stringFromDialogResult:systemDialogResult];
        text = [NSString stringWithFormat:@"User Action: %@\nSystem Action: %@", userResultString, systemResultString];
    }
    resultLabel.text = text;

    [UIView animateWithDuration:0.35 animations:^{
        resultLabel.alpha = 1.0;
    }];
}

- (NSString *) stringFromDialogResult:(ClusterDialogResult)result
{
    switch (result) {
        case ClusterDialogResultNoActionTaken:
            return @"No Action Taken";
            break;
        case ClusterDialogResultGranted:
            return @"Granted";
            break;
        case ClusterDialogResultDenied:
            return @"Denied";
            break;
        case ClusterDialogResultParentallyRestricted:
            return @"Restricted";
            break;
        default:
            break;
    }
}
@end
